src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java

```diff
@@ -0,0 +1,189 @@
+ */
+@Service
+public class LearningGoalProgressService {
+
+    private final Logger logger = LoggerFactory.getLogger(LearningGoalProgressService.class);
+
+    private final LearningGoalRepository learningGoalRepository;
+
+    private final LearningGoalProgressRepository learningGoalProgressRepository;
+
+    private final StudentScoreRepository studentScoreRepository;
+
+    private final TeamScoreRepository teamScoreRepository;
+
+    private final ExerciseRepository exerciseRepository;
+
+    private final LectureUnitRepository lectureUnitRepository;
+
+    public LearningGoalProgressService(LearningGoalRepository learningGoalRepository, LearningGoalProgressRepository learningGoalProgressRepository,
+            StudentScoreRepository studentScoreRepository, TeamScoreRepository teamScoreRepository, ExerciseRepository exerciseRepository,
+            LectureUnitRepository lectureUnitRepository) {
+        this.learningGoalRepository = learningGoalRepository;
+        this.learningGoalProgressRepository = learningGoalProgressRepository;
+        this.studentScoreRepository = studentScoreRepository;
+        this.teamScoreRepository = teamScoreRepository;
+        this.exerciseRepository = exerciseRepository;
+        this.lectureUnitRepository = lectureUnitRepository;
+    }
+
+    /**
+     * Asynchronously update the progress for all learning goals linked to the given learning object
+     * @param learningObject The learning object for which to fetch the learning goals
+     * @param participant The participant (user or team) for which to update the progress
+     */
+    @Async
+    public void updateProgressByLearningObjectAsync(ILearningObject learningObject, @NotNull Participant participant) {
+        updateProgressByLearningObject(learningObject, participant);
+    }
+
+    /**
+     * Update the progress for all learning goals linked to the given learning object
+     * @param learningObject The learning object for which to fetch the learning goals
+     * @param participant The participant (user or team) for which to update the progress
+     */
+    public void updateProgressByLearningObject(ILearningObject learningObject, @NotNull Participant participant) {
+        logger.debug("Updating learning goal progress for participant {}.", participant.getName());
+        try {
+            SecurityUtils.setAuthorizationObject();
+
+            Set<LearningGoal> learningGoals;
+            if (learningObject instanceof Exercise exercise) {
+                learningGoals = exerciseRepository.findByIdWithLearningGoals(exercise.getId()).map(Exercise::getLearningGoals).orElse(null);
+            }
+            else if (learningObject instanceof LectureUnit lectureUnit) {
+                learningGoals = lectureUnitRepository.findByIdWithLearningGoals(lectureUnit.getId()).map(LectureUnit::getLearningGoals).orElse(null);
+            }
+            else {
+                return;
```

Comment:
Not necessary but down below you throw an exception such aus:
```
 throw new IllegalArgumentException("Completable must be either LectureUnit or Exercise");
``` 
Maybe do it here to?